community/rfcs/25-03-14-GenAIExample-001-CodeTrans-with-Agents.md

```diff
@@ -0,0 +1,346 @@
+
+    ```python
+    import ast
+
+    code = "xxx"
+    tree = ast.parse(code)
+    # analyze each node in ast tree
+    for node in ast.walk(tree):
+        # Do import checks
+    ```
+* Install dependencies automatically
+
+  * Code execution tool need to support extracting and installing dependencies from source code automatically.
+  * Use AST here to extract the `import`/`from xxx import xxx` libraries.
+
+    ```python
+    import ast
+
+    def extract_imports(code_str):
+        tree = ast.parse(code_str)
+        imports = set()
+        for node in ast.walk(tree):
+            if isinstance(node, ast.Import):
+                for alias in node.names:
+                    imports.add(alias.name.split('.')[0])
+            elif isinstance(node, ast.ImportFrom) and node.module:
+                imports.add(node.module.split('.')[0])
+        return list(imports)
+    ```
+* Sandbox execution
+
+  * To ensure that the code runs in a fully isolated environment, the tool needs to use container-based sandbox like `Docker`.
+  * In the context of resource constraints, the resource limits (including memory, CPU, process numbers) and security policies are needed.
+  * Since the dependencies need to be installed first, the network authority will be processed in two stages:
+    1. preperation: install the dependencies into a mounted path
+       ```bash
+       # add memory, cpu, process limit, and disable privilege operations
+       docker run --rm -m 512m --cpus="1" --pids-limit=128 \
+         --security-opt no-new-privileges -v $(pwd)/code:/code \
+         sandbox-python:3.10 \
+         bash -c "pip install -r /code/requirements.txt -t /code/.deps"
+       ```
+    2. execution: run codes using mounted dependencies
+       ```bash
+       docker run --rm -v $(pwd):/code --network=none \
+         sandbox-python:3.10 \
+         python3 -I -E -S /code/user_code.py
+       ```
+
+## Expected Benefits
+
+| Feature                      | Benefits                                              |
+| ---------------------------- | ----------------------------------------------------- |
+| Input Code Validation        | Catches errors early, preventing faulty translations. |
+| Output Code Validation       | Ensures reliable and accurate code conversion.        |
+| Automated Debug Feedback     | Reduces trial-and-error, improving LLM accuracy.      |
+| Lint Static Code Check       | Catch bugs early and enforce consistent code quality. |
+| Secure Execution Environment | Protects the system from malicious code.              |
+| Error Classification         | Identifies syntax, logic errors for better debugging. |
+
+## Implementation Plan
+
+### Phase 1: Develop Code Execution Tool, target v1.3
+
+- Develop the Code Execution Tool in Agent to provide a secure execution environment.
+- Ensure the code execution environment is securely isolated.
```

Comment:
IMHO this is less important that the linting features, intended to cover just one of the languages, and more likely to be service security and performance issue, so it should be done later.